README.md
# chat_bubble

A chat bubble widget, similar to WeChat, You can customize the orientation and the size of the rounded corners, as well as the rounded corners and position of the triangle..

类似微信一样的聊天气泡，可以用作文字、图片等等。 通常文字类的聊天气泡可以使用定位的纯色的小矩形旋转，但把它应用在图片上面就不对了。这个插件可以在图片左右侧切割出一个气泡小三角。

## Depend on it
Add this to your package's pubspec.yaml file:
```
dependencies:
  chat_bubble: any
```

## Usage
```
import 'package:chat_bubble/chat_bubble.dart';


ChatBubble(
  // 阴影
  shadows: <Shadow>[
    Shadow(
      color: Colors.grey,
      offset: Offset(-1.0, -1.0),
      blurRadius: 0.1,
    ),
    Shadow(
      color: Colors.grey,
      offset: Offset(1.0, 1.0),
      blurRadius: 0.4,
    )
  ],
  // 小三角的方向
  direction: reverse ? ChatBubbleNipDirection.RIGHT:  ChatBubbleNipDirection.LEFT,
  // 小三角的边长
  // nipLength: 30.0,
  // 小三角的圆角
  // nipRadius: 10.0,
  // 小三角到顶部的距离
  // nipTop: 80.0,
  child: Container(
    color: Colors.grey,
    child: img,
  ),
)
```

## Screenshots

![截图](https://github.com/mrTangYun/flutter_chat_bubble/raw/master/screenshots/Screenshot_1587114817.png)

![说明](https://github.com/mrTangYun/flutter_chat_bubble/raw/master/screenshots/Screenshot_1587095359.png)


CHANGELOG.md
## [0.1.1+4]
项目说明

test/chat_bubble_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:chat_bubble/chat_bubble.dart';
import 'package:flutter/material.dart';

void main() {
  test('adds one to input values', () {
     ChatBubble(
      child: Text('test')
    );
    // final calculator = Calculator();
    // expect(calculator.addOne(2), 3);
    // expect(calculator.addOne(-7), -6);
    // expect(calculator.addOne(0), 1);
    // expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


lib/bubble.dart
library chat_bubble;

import 'dart:math';
import 'package:flutter/material.dart';

abstract class Bubble extends CustomClipper<Path> {
  /// 圆角大小
  double radius;

  /// 三角形开始的位置
  double triangleTop;

  /// 三角形边长
  double triangleLength;

  /// 三角形圆角大小
  double triangleRadius;

  /// 三角形高度的一半
  double _triangleHalfHeight;

  /// 三角形圆角的高度(宽度)
  double _triangleRadiusHeight;

  Bubble(
      {this.radius, this.triangleTop, this.triangleLength, this.triangleRadius})
      : this._triangleHalfHeight = sqrt(2 * pow(triangleLength, 2)) / 2,
        this._triangleRadiusHeight = cos(pi * 45 / 180) * triangleRadius;
}

class BubbleClipperRight extends Bubble {
  BubbleClipperRight({radius, triangleTop, triangleLength, triangleRadius})
      : super(
          radius: radius,
          triangleTop: triangleTop,
          triangleLength: triangleLength,
          triangleRadius: triangleRadius,
        );
  @override
  Path getClip(Size size) {
    final Path path = Path();
    path.lineTo(0 + this.radius, 0);

    // 右上角
    path.lineTo(size.width - this._triangleHalfHeight - this.radius, 0);
    final Offset _pc1 = Offset(size.width - this._triangleHalfHeight, 0);
    final Offset _pe1 =
        Offset(size.width - this._triangleHalfHeight, this.radius);
    path.quadraticBezierTo(_pc1.dx, _pc1.dy, _pe1.dx, _pe1.dy);
    path.lineTo(_pe1.dx, _pe1.dy);

    // 三角形
    path.lineTo(_pe1.dx, this.triangleTop);
    path.lineTo(
        size.width - this._triangleRadiusHeight,
        this.triangleTop +
            this._triangleHalfHeight -
            this._triangleRadiusHeight);
    final Offset _pct =
        Offset(size.width, this.triangleTop + this._triangleHalfHeight);
    final Offset _pet = Offset(
        size.width - this._triangleRadiusHeight,
        this.triangleTop +
            this._triangleHalfHeight +
            this._triangleRadiusHeight);
    path.quadraticBezierTo(_pct.dx, _pct.dy, _pet.dx, _pet.dy);
    path.lineTo(_pet.dx, _pet.dy);
    path.lineTo(_pe1.dx, this.triangleTop + this._triangleHalfHeight * 2);

    // 右下角
    path.lineTo(_pe1.dx, size.height - this.radius);
    final Offset _pc2 = Offset(_pe1.dx, size.height);
    final Offset _pe2 = Offset(_pe1.dx - this.radius, size.height);
    path.quadraticBezierTo(_pc2.dx, _pc2.dy, _pe2.dx, _pe2.dy);
    path.lineTo(_pe2.dx, _pe2.dy);

    // 左下角
    path.lineTo(0 + this.radius, size.height);
    final Offset _pc3 = Offset(0, size.height);
    final Offset _pe3 = Offset(0, size.height - this.radius);
    path.quadraticBezierTo(_pc3.dx, _pc3.dy, _pe3.dx, _pe3.dy);
    path.lineTo(_pe3.dx, _pe3.dy);

    // 左上角
    path.lineTo(0, 0 + this.radius);
    final Offset _pc4 = Offset(0, 0);
    final Offset _pe4 = Offset(0 + this.radius, 0);
    path.quadraticBezierTo(_pc4.dx, _pc4.dy, _pe4.dx, _pe4.dy);
    path.lineTo(_pe4.dx, _pe4.dy);
    return path;
  }

  @override
  bool shouldReclip(CustomClipper<Path> oldClipper) {
    return false;
  }
}

class BubbleClipperLeft extends Bubble {
  BubbleClipperLeft({radius, triangleTop, triangleLength, triangleRadius})
      : super(
          radius: radius,
          triangleTop: triangleTop,
          triangleLength: triangleLength,
          triangleRadius: triangleRadius,
        );
  @override
  Path getClip(Size size) {
    final Path path = Path();
    path.lineTo(0 + this.radius + this._triangleHalfHeight, 0);

    // 右上角
    path.lineTo(size.width - this.radius, 0);
    final Offset _pc1 = Offset(size.width, 0);
    final Offset _pe1 = Offset(size.width, this.radius);
    path.quadraticBezierTo(_pc1.dx, _pc1.dy, _pe1.dx, _pe1.dy);
    path.lineTo(_pe1.dx, _pe1.dy);

    // 右下角
    path.lineTo(_pe1.dx, size.height - this.radius);
    final Offset _pc2 = Offset(_pe1.dx, size.height);
    final Offset _pe2 = Offset(_pe1.dx - this.radius, size.height);
    path.quadraticBezierTo(_pc2.dx, _pc2.dy, _pe2.dx, _pe2.dy);
    path.lineTo(_pe2.dx, _pe2.dy);

    // 左下角
    path.lineTo(0 + this.radius + this._triangleHalfHeight, size.height);
    final Offset _pc3 = Offset(0 + this._triangleHalfHeight, size.height);
    final Offset _pe3 =
        Offset(0 + this._triangleHalfHeight, size.height - this.radius);
    path.quadraticBezierTo(_pc3.dx, _pc3.dy, _pe3.dx, _pe3.dy);
    path.lineTo(_pe3.dx, _pe3.dy);

    // 三角形
    path.lineTo(_pe3.dx, this.triangleTop + this._triangleHalfHeight * 2);
    path.lineTo(
        0 + this._triangleRadiusHeight,
        this.triangleTop +
            this._triangleHalfHeight +
            this._triangleRadiusHeight);
    final Offset _pct = Offset(0, this.triangleTop + this._triangleHalfHeight);
    final Offset _pet = Offset(
        0 + this._triangleRadiusHeight,
        this.triangleTop +
            this._triangleHalfHeight -
            this._triangleRadiusHeight);
    path.quadraticBezierTo(_pct.dx, _pct.dy, _pet.dx, _pet.dy);
    path.lineTo(_pet.dx, _pet.dy);
    path.lineTo(_pe3.dx, this.triangleTop);

    // 左上角
    path.lineTo(_pe3.dx, 0 + this.radius);
    final Offset _pc4 = Offset(_pe3.dx, 0);
    final Offset _pe4 = Offset(_pe3.dx + this.radius, 0);
    path.quadraticBezierTo(_pc4.dx, _pc4.dy, _pe4.dx, _pe4.dy);
    path.lineTo(_pe4.dx, _pe4.dy);
    return path;
  }

  @override
  bool shouldReclip(CustomClipper<Path> oldClipper) {
    return false;
  }
}


lib/chat_bubble.dart
library chat_bubble;

import 'package:chat_bubble/bubble.dart';
import 'package:chat_bubble/clipShadowPath.dart';
import 'package:flutter/material.dart';

enum ChatBubbleNipDirection { LEFT, RIGHT }

class ChatBubble extends StatelessWidget {
  /// 圆角大小
  final double radius;

  /// 三角形距离顶部的距离
  final double nipTop;

  /// 三角形边长
  final double nipLength;

  /// 三角形圆角大小
  final double nipRadius;

  final Widget child;
  final List<Shadow> shadows;
  final ChatBubbleNipDirection direction;
  ChatBubble({
    @required this.child,
    this.shadows,
    this.direction = ChatBubbleNipDirection.RIGHT,
    this.radius = 6.0,
    this.nipTop = 15.0,
    this.nipLength = 10.0,
    this.nipRadius = 2.0,
  });
  @override
  Widget build(BuildContext context) {
    return ClipShadowPath(
      shadows: this.shadows,
      clipper: this.direction == ChatBubbleNipDirection.RIGHT
          ? BubbleClipperRight(
              radius: radius,
              triangleTop: nipTop,
              triangleLength: nipLength,
              triangleRadius: nipRadius,
            )
          : BubbleClipperLeft(
              radius: radius,
              triangleTop: nipTop,
              triangleLength: nipLength,
              triangleRadius: nipRadius,
            ),
      child: child,
    );
  }
}


lib/clipShadowPath.dart
import 'package:flutter/material.dart';

@immutable
class ClipShadowPath extends StatelessWidget {
  final List<Shadow> shadows;
  final CustomClipper<Path> clipper;
  final Widget child;

  ClipShadowPath({
    this.shadows,
    @required this.clipper,
    @required this.child,
  });

  @override
  Widget build(BuildContext context) {
    if (this.shadows != null && this.shadows.length > 0) {
      return CustomPaint(
        painter: _ClipShadowShadowPainter(
          clipper: this.clipper,
          shadows: this.shadows,
        ),
        child: ClipPath(child: child, clipper: this.clipper),
      );
    } else {
      return ClipPath(child: child, clipper: this.clipper);
    }
  }
}

class _ClipShadowShadowPainter extends CustomPainter {
  final List<Shadow> shadows;
  final CustomClipper<Path> clipper;

  _ClipShadowShadowPainter({@required this.shadows, @required this.clipper});

  @override
  void paint(Canvas canvas, Size size) {
    if (shadows.length > 0) {
      for (var i = 0; i < shadows.length; i++) {
        Shadow shadow = shadows[i];
        var paint = shadow.toPaint();
        var clipPath = clipper.getClip(size).shift(shadow.offset);
        canvas.drawPath(clipPath, paint);
      }
    }
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return false;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:chat_bubble/chat_bubble.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatelessWidget {
  final String title;
  MyHomePage({Key key, this.title}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    bool reverse = true;
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: SingleChildScrollView(
          child: Padding(
            padding: const EdgeInsets.all(8.0),
            child: Column(
              // Column is also a layout widget. It takes a list of children and
              // arranges them vertically. By default, it sizes itself to fit its
              // children horizontally, and tries to be as tall as its parent.
              //
              // Invoke "debug painting" (press "p" in the console, choose the
              // "Toggle Debug Paint" action from the Flutter Inspector in Android
              // Studio, or the "Toggle Debug Paint" command in Visual Studio Code)
              // to see the wireframe for each widget.
              //
              // Column has various properties to control how it sizes itself and
              // how it positions its children. Here we use mainAxisAlignment to
              // center the children vertically; the main axis here is the vertical
              // axis because Columns are vertical (the cross axis would be
              // horizontal).
              mainAxisAlignment: MainAxisAlignment.center,
              children: <Widget>[
                ChatBubble(
                  direction: ChatBubbleNipDirection.LEFT,
                  child: Container(
                    height: 45,
                    color: Color(0xffdddddd),
                    child: Center(
                      child: Text('收到文本'),
                    ),
                  ),
                ),
                SizedBox(
                  height: 30,
                ),
                ChatBubble(
                  direction: ChatBubbleNipDirection.RIGHT,
                  child: Container(
                    width: 230,
                    height: 45,
                    color: Color(0xffa5ed7e),
                    child: Center(
                      child: Text('发送文本'),
                    ),
                  ),
                ),
                SizedBox(
                  height: 30,
                ),
                Align(
                  alignment: Alignment.bottomRight,
                  child: ChatBubble(
                    shadows: <Shadow>[
                      Shadow(
                        color: Colors.red,
                        offset: Offset(0, 0),
                        blurRadius: 5,
                      ),
                    ],
                    direction: ChatBubbleNipDirection.RIGHT,
                    child: Container(
                      width: 230,
                      height: 45,
                      color: Color(0xffa5ed7e),
                      child: Center(
                        child: Text('单个阴影'),
                      ),
                    ),
                  ),
                ),
                SizedBox(
                  height: 30,
                ),
                ChatBubble(
                  shadows: <Shadow>[
                    Shadow(
                      color: Colors.red,
                      offset: Offset(-5, -5),
                      blurRadius: 5,
                    ),
                    Shadow(
                      color: Colors.blue,
                      offset: Offset(5, 5),
                      blurRadius: 5,
                    )
                  ],
                  direction: ChatBubbleNipDirection.RIGHT,
                  child: Container(
                    width: 230,
                    height: 45,
                    color: Color(0xffa5ed7e),
                    child: Center(
                      child: Text('多个阴影'),
                    ),
                  ),
                ),
                SizedBox(
                  height: 30,
                ),
                ChatBubble(
                  shadows: <Shadow>[
                    Shadow(
                      color: Colors.red,
                      offset: Offset(0, 0),
                      blurRadius: 5,
                    )
                  ],
                  direction: reverse
                      ? ChatBubbleNipDirection.RIGHT
                      : ChatBubbleNipDirection.LEFT,
                  child: Container(
                    width: 300,
                    color: Colors.grey.withOpacity(0.8),
                    child: Image.network(
                      'https://pics7.baidu.com/feed/b21c8701a18b87d6627be358ca93ae3e1e30fdda.jpeg?token=a6434449ed9bc79aa3c414a52e4a4e1b',
                    ),
                  ),
                ),
                SizedBox(
                  height: 30,
                ),
                ChatBubble(
                  nipLength: 30.0,
                  nipRadius: 10.0,
                  nipTop: 80.0,
                  shadows: <Shadow>[
                    Shadow(
                      color: Colors.red,
                      offset: Offset(0, 0),
                      blurRadius: 5,
                    )
                  ],
                  direction: reverse
                      ? ChatBubbleNipDirection.RIGHT
                      : ChatBubbleNipDirection.LEFT,
                  child: Container(
                    width: 300,
                    color: Colors.grey.withOpacity(0.8),
                    child: Image.network(
                      'https://pics7.baidu.com/feed/b21c8701a18b87d6627be358ca93ae3e1e30fdda.jpeg?token=a6434449ed9bc79aa3c414a52e4a4e1b',
                    ),
                  ),
                ),
                SizedBox(
                  height: 30,
                ),
                ChatBubble(
                  shadows: <Shadow>[
                    Shadow(
                      color: Colors.grey.withOpacity(0.8),
                      offset: Offset(-1.0, -1.0),
                      blurRadius: 0.1,
                    ),
                    Shadow(
                      color: Colors.grey.withOpacity(0.8),
                      offset: Offset(1.0, 1.0),
                      blurRadius: 0.4,
                    )
                  ],
                  direction: ChatBubbleNipDirection.LEFT,
                  child: Container(
                    width: 200,
                    color: Colors.grey.withOpacity(0.8),
                    child: Image.network(
                      'https://pics7.baidu.com/feed/b21c8701a18b87d6627be358ca93ae3e1e30fdda.jpeg?token=a6434449ed9bc79aa3c414a52e4a4e1b',
                    ),
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
